src/include/main/db_main.h

```diff
@@ -102,37 +58,648 @@ class DBMain {
+      }
+
+      // Run the GC to get a clean system. This needs to be done before instantiating the GC thread
+      // because the GC is not thread-safe
+      deferred_action_manager_->FullyPerformGC(garbage_collector_, log_manager_);
+    }
+
+    ~CatalogLayer() {
+      catalog_->TearDown();  // generates txns and deferred actions, so need to flush the system afterwards
+      deferred_action_manager_->FullyPerformGC(common::ManagedPointer(garbage_collector_), log_manager_);
+    }
+
+    /**
+     * @return ManagedPointer to the component
+     */
+    common::ManagedPointer<catalog::Catalog> GetCatalog() const { return common::ManagedPointer(catalog_); }
+
+   private:
+    const common::ManagedPointer<transaction::DeferredActionManager> deferred_action_manager_;
+    const common::ManagedPointer<storage::GarbageCollector> garbage_collector_;
+    const common::ManagedPointer<storage::LogManager> log_manager_;
+
+    std::unique_ptr<catalog::Catalog> catalog_;
+  };
 
   /**
-   * Cleans up and exit.
+   * ConnectionHandleFactory, CommandFactory, ProtocolInterpreter::Provider, Server
    */
-  void CleanUp();
+  class NetworkLayer {
+   public:
+    /**
+     * @param thread_registry argument to the TerrierServer
+     * @param traffic_cop argument to the ConnectionHandleFactor
+     * @param port needed for safe destruction of CatalogLayer if logging is enabled
+     */
+    NetworkLayer(const common::ManagedPointer<common::DedicatedThreadRegistry> thread_registry,
+                 const common::ManagedPointer<trafficcop::TrafficCop> traffic_cop, const uint16_t port) {
+      connection_handle_factory_ = std::make_unique<network::ConnectionHandleFactory>(traffic_cop);
+      command_factory_ = std::make_unique<network::PostgresCommandFactory>();
+      provider_ =
+          std::make_unique<network::PostgresProtocolInterpreter::Provider>(common::ManagedPointer(command_factory_));
+      server_ = std::make_unique<network::TerrierServer>(
+          common::ManagedPointer(provider_), common::ManagedPointer(connection_handle_factory_), thread_registry, port);
+    }
+
+    /**
+     * @return ManagedPointer to the component
+     */
+    common::ManagedPointer<network::TerrierServer> GetServer() const { return common::ManagedPointer(server_); }
+
+   private:
+    // Order matters here for destruction order
+    std::unique_ptr<network::ConnectionHandleFactory> connection_handle_factory_;
+    std::unique_ptr<network::PostgresCommandFactory> command_factory_;
+    std::unique_ptr<network::ProtocolInterpreter::Provider> provider_;
+    std::unique_ptr<network::TerrierServer> server_;
+  };
+
+  /**
+   * Creates DBMain objects
+   */
+  class Builder {
+   public:
+    /**
+     * Validate configuration and construct requested DBMain.
+     * @return DBMain that you get to own. YES, YOU! We're just giving away DBMains over here!
+     */
+    std::unique_ptr<DBMain> Build() {
+      auto db_main = std::make_unique<DBMain>();
+
```

Comment:
Do you think it is worth having a high level comment on dependencies here? Or maybe a link to a wiki page that has the graph which you drew on the board.

I _think_ we add new components infrequently enough that this shouldn't fall too far out of date / will remain quite useful overall? Defer to your judgment.

Reply:
I could document that here, yeah. I considered writing a lot more about how to add a dependency to `DBMain` but I'm hoping in the next 2 weeks everything is in anyway.